internal/providers/zvm/zvm.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The zVM provider fetches a local configuration from the virtual unit
+// record devices.
+
+package zvm
+
+import (
+	"io/ioutil"
+	"os/exec"
+	"strings"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/coreos/ignition/v2/internal/distro"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/providers/util"
+	"github.com/coreos/ignition/v2/internal/resource"
+	"github.com/coreos/vcontext/report"
+)
+
+const readerDevice string = "000c"
+const configFile string = "config.ign"
+
+func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	// Fetch config files directly from reader device.
+	_, err := f.Logger.LogCmd(exec.Command(distro.ModprobeCmd(), "vmur"), "Loading zVM control program module")
+	if err != nil {
+		f.Logger.Err("Couldn't install vmur module: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	// Online the reader device.
+	logger := f.Logger
+	err = onlineDevice(logger)
+	if err != nil {
+		return types.Config{}, report.Report{}, err
+	}
+	// Read files from the z/VM reader queue.
+	readerInfo, err := exec.Command(distro.VmurCmd(), "li").CombinedOutput()
+	if err != nil {
+		f.Logger.Err("Can not get reader device: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	items := strings.Split(string(readerInfo), "\n")
+	for _, records := range items {
+		record := strings.Fields(records)
+		if len(record) <= 1 {
+			break
+		}
+		spoolid := record[1]
+		file := record[9]
+		ftype := record[10]
+		// Receive the spool file.
+		if file == "ignition" {
+			if ftype != "" {
+				file = record[9] + "." + ftype
+			}
+			_, err := f.Logger.LogCmd(exec.Command(distro.VmurCmd(), "re", "-f", spoolid, file), "Receive the spool file")
+			if err == nil {
+				f.Logger.Info("using config file at %q", file)
+				_, err := f.Logger.LogCmd(exec.Command(distro.TarCmd(), "-xzvf", file), "Extract the spool file")
```

Comment:
You previously replied:
> Sorry. I forgot this question. I mean If not using tarball, vmur will format the json config file with padding char. There will error when do later ParseConfig operation. So to avoid the config file being formatted , I chose to use tarball.

What is padding char in this case? If the config is just padded I think it would be better to just strip the padding than to need to run tar?

Reply:
@ajeddeloh Hi, The process is :
Write data to the z/VM punch file queue and transfer it to virtual reader. The data is sliced up into 80-byte or 132-byte chunks (called records) and written to the punch device. If the data length is not an integer multiple of 80 or 132, the last record is padded. So The length of padded string is not confirmed.

Reply:
@ajeddeloh I have resolved the padding string with bytes package. So now tar is not necessary.